### ISSUE TITLE
[New Concept Exercise] :  dict-methods

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2348]
This issue describes how to implement the `dict-methods` (dictionary methods) concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/use of useful `dict-methods`  (methods from the `dict` - dictionary class) and some additional practice/techniques for working with `dicts`.

## Learning objectives

Cover useful `dict` methods and a few techniques for operating on/manipulating `dicts`.

- `dict` methods :

  -  `get()` 
  -  `clear()`
  - `popitem()` 
  -  `update()` 
  - *classmethod*[ `fromkeys`(*iterable*[, *value*])](https://docs.python.org/3/library/stdtypes.html#dict.fromkeys)
  - Pythong 3.9 pipe or union operator (`d | other`, `d |= other`) - **note:** may not be included.  Decision still TBD.

- Working more with  the `dict` views `items()` , `keys()` or `values()`.  (e.g, by sorting information using `sorted()` or by swapping `keys` and `values`, etc.)
- Knowing that Dictionaries can be _nested_, _-- e.g._ ' a dictionary of dictionaries'.
- Considerations when `updating()` or using `union` with dictionaries.

  

## Out of scope

Please take a look at the `dicts` concept exercise [design.md file](https://github.com/exercism/python/edit/main/exercises/concept/inventory-management/.meta/design.md) for `dict` features  taught thus far.  While those methods can be used for solutions to this exercise, it  isn't necessary to cover them again in detail.  Additionally, the following is out of scope:

- Dictionary comprehensions
- Built-in functions as they relate to this data structure (*e.g.* `len()`, or `enumerate()`
- Considerations of Mutability
- `copy()` vs `deepcopy()`
- Memory and performance characteristics.
- Related `collections` module with `Counter()` and `defaultdict()`

## Concepts

- `dicts`
- `dict-methods`

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `bools`
- `conditionals`
- `comparisons`
- `dicts`
- `lists`
- `loops`
- `numbers`
- `strings`
- `tuples`

## Resources to refer to

- [Python docs:  Tutorial - Dictionaries](https://docs.python.org/3/tutorial/datastructures.html#dictionaries)
- [Python docs:  Mapping Type `dict`](https://docs.python.org/3/library/stdtypes.html#mapping-types-dict)
- [Real Python:  Dicts](https://realpython.com/python-dicts/)
- [Digital Ocean:  Understanding dictionaries in python 3](https://www.digitalocean.com/community/tutorials/understanding-dictionaries-in-python-3)
- [Stack Overflow: exchanging keys with values in a `dict` in Python](https://stackoverflow.com/questions/1031851/how-do-i-exchange-keys-with-values-in-a-dictionary)
- [kite:  how to sort a dictionary by key in python](https://www.kite.com/python/answers/how-to-sort-a-dictionary-by-key-in-python)
- [medium:  16 Python Dictionary Tips](https://medium.com/python-in-plain-english/16-intermediate-level-python-dictionary-tips-tricks-and-shortcuts-1376859e1adc)  _**note:** this is a good resource for ideas and writing this exericse, but is a subscription-based service, so not the best for linking to_

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  * You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  *  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  *  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson) & [concept exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md)

* ### Concept `about.md`

  **Concept file/issue**: There is a PR pending for this concept that includes `about.md`, `introduction.md`, and `links.json` :   [PR 2374](https://github.com/exercism/python/pull/2347).  If the PR has been merged, the files can be found under `python/concepts/dict-methods/`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  * This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

  
## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

* **Exercise UUID** : `1f67781e-07a3-4d00-a49f-59493e7b7f2c`

* **concepts** should be filled in from the Concepts section in this issue
* **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

* `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.

* `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.

In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.

While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.

Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
Flagging as a close and re-open with new issue template.
